File: data_structures/QuickSort.java
import java.util.Arrays;

class QuickSort {
    public int partition(int[] arr, int start, int end) {
        int pivot = arr[end];
        int pIndex = start;
        for(int i = start; i < end; i++) {
            if(arr[i] <= pivot) {
                swap(arr[i], arr[pIndex]);
                pIndex++;
            }
        }
        swap(arr[pIndex], arr[end]);
        return pIndex;
    }

    public int[] sort(int[] arr, int start, int end) {
        if(start < end) {
            int partitionIndex = partition(arr, start, end);
            sort(arr, start, partitionIndex-1);
            sort(arr, partitionIndex+1, end);
        }
        return arr;
    }    

    public void swap(int low, int high) {
        int temp = low;
        low = high;
        high = temp;
    }

    public static void main(String[] args) {
        QuickSort obj = new QuickSort();
        int[] A = {7, 6, 5, 4, 3, 2, 1, 0};
        System.out.println(Arrays.toString(obj.sort(A, 0, 7)));
    }
}

File: algorithms/sort/quick_sort/kotlin/quicksort.kt
class QuickSort {
    private fun quickSort(arr: IntArray, start: Int, end: Int) {
        if (start < end) {
            val pivot_pos = partition(arr, start, end)
            quickSort(arr, start, pivot_pos - 1)
            quickSort(arr, pivot_pos + 1, end)
        }
    }

    private fun partition(arr: IntArray, start: Int, end: Int): Int {
        val pivot = arr[end]
        var i = start
        var j = end - 1
        while (i <= j) {
            if (arr[j] <= pivot && arr[i] > pivot) {
                swap(arr, i, j)
            }
            if (arr[i] <= pivot) {
                i++
            }
            if (arr[j] > pivot) {
                j--
            }
        }
        swap(arr, i, end)
        return i
    }

    private fun swap(arr: IntArray, i: Int, j: Int) {
        val tmp = arr[i]
        arr[i] = arr[j]
        arr[j] = tmp
    }

    private fun print(arr: IntArray) {
        for (i in arr.indices) {
            print(arr[i].toString() + " ")
        }
        print('\n')
    }

    fun main() {
        val arr = intArrayOf(9, 7, 5, 11, 12, 2, 14, 3, 10, 4, 6)
        print("Before: ")
        print(arr)
        quickSort(arr, 0, arr.size - 1)
        print("After: ")
        print(arr)
    }
}
